File: 프로그래머스/lv1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.java
import java.util.Arrays;
class Solution {
    public int solution(String dartResult) {
        int answer = 0;

        int[] result = new int[3];
        int n=0;
        int idx=0;
        String num="";      // 숫자를 저장하는 문자열
        for(int i=0; i<dartResult.length(); i++){
            char c = dartResult.charAt(i);

            if(Character.isDigit(c)){ 
                num += String.valueOf(c);
            }
            
            if(c=='S'||c=='D'||c=='T'){
                n=Integer.parseInt(num);
                if(c=='S'){
                    result[idx++]=n;
                }
                else if(c=='D'){
                    result[idx++]=(int)Math.pow(n,2);
                }
                else{
                    result[idx++]=(int)Math.pow(n,3);
                }
                num="";     // 문자열 초기화, 다음 수를 저장하기 위해
            }
            
            // 옵션 어려움
            else {
                if(c=='*'){
                    result[idx-1]*=2;
                    if(idx >= 2) {
                        result[idx-2]*=2; 
                    }
                }
                if(c=='#') {
                    result[idx-1]*=(-1);
                }
            }
        }

        for(int i=0; i<3; i++){
            answer += result[i];
        }
        return answer;
    }
}

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    answer = 0
    
    num_list = [0]*3       # 합계 계산
    nums = ""              # 점수에서 숫자를 저장
    idx = 0                # num_list의 인덱스
    for i in range(len(dartResult)):
        r = dartResult[i]
        # nums에 숫자 저장
        # nums가 문자열인 이유는, 10을 저장하기 위해서
        if r.isdigit():
            nums += r
        if r == '#': 
            num_list[idx-1] *= -1      # 현재 점수에 마이너스
        if r == '*':
            num_list[idx-1] *= 2       # 현재 점수에 두배
            num_list[idx-2] *= 2   # 이전 점수에 두배
                
        if r == 'S':
            num_list[idx] = int(nums)
            idx += 1          # 인덱스 증가
            nums = ""         # nums 초기화
        if r == 'D':
            num_list[idx] = int(nums)**2
            idx += 1
            nums = ""
        if r == 'T':
            num_list[idx] = int(nums)**3
            idx += 1
            nums = ""

    return sum(num_list)